dask/dataframe/core.py

```diff
@@ -4422,6 +4743,15 @@ def iloc(self):
         return _iLocIndexer(self)
 
     def __len__(self):
+        try:
+            # Use partition statistics if they are available
+            part_sizes = self.partition_metadata.get_stats({"__num_rows__"})[
+                "__num_rows__"
+            ]
+            if part_sizes:
+                return sum(part_sizes)
+        except KeyError:
+            pass
```

Comment:
I think this should just be an optional top-level attribute on the metadata, rather than keying into a dictionary with a magic name.